### ISSUE TITLE
Boolean population not correctly loaded

### ISSUE THREAD

**Issue by attaxia** [AmpersandTarski/RAP#117]
#### De situatie
Boolean populations are not loaded as specified

#### Wat er gebeurde er en wat had je verwacht?
A population for a relation with a boolean atom was specified, but the correct values are not loaded in the prototype

#### Stappen om dit te reproduceren
1. Load script provided
2. Generate prototype
3. Go to "Artikels" interface
4. Click "artikel1" which is of "artikeltype1"
5. Click "artikeltype1"
5. Checkbox is unchecked for "inleverplichtig", although it should be true according to provided population
6. By clicking the artikeltype1 the checkbox can be checked/unchecked, this value is properly stored and then shown correctly in the Artikel interface

#### Screenshot / Video
<img width="450" alt="Screen Shot 2021-01-05 at 12 45 05" src="https://user-images.githubusercontent.com/4743303/103642733-e4891100-4f53-11eb-906a-33c29339c676.png">

#### Context / Source van ampersand script

```
CONTEXT Loonbewerking IN ENGLISH

CONCEPT Artikel "Een artikel"
CONCEPT Artikeltype "Artikeltype"
CONCEPT Inleverplichtig "Inleverplichtig"
REPRESENT Inleverplichtig TYPE BOOLEAN

RELATION artikelType [Artikel*Artikeltype]              [UNI, TOT]

RELATION inleverPlichtig [Artikeltype*Inleverplichtig]  [UNI, TOT]


POPULATION artikelType CONTAINS
[ ("artikel1", "artikeltype1")
, ("artikel2", "artikeltype2")
, ("artikel3", "artikeltype2")
]


POPULATION inleverPlichtig CONTAINS
[ ("artikeltype1", TRUE)
, ("artikeltype2", FALSE)
]

INTERFACE Artikels : "_SESSION"                                 cRud
BOX <TABS>
     [ "Alle artikels" : V[SESSION*Artikel]                     CRud
       BOX <TABLE>
                [ "Artikel" : I                                 cRud
                ]
     ]

INTERFACE Artikel FOR Artikel : I[Artikel]
BOX [ "Artikel identifier" : I[Artikel]
    , "Artikeltype" : artikelType     
    ]

INTERFACE Artikeltype FOR Artikeltype : I[Artikeltype]
BOX [ "Artikeltype" : I[Artikeltype]
    , "Inleverplichtig" : inleverPlichtig           
    ]

     
ENDCONTEXT
```

**Comment by stefjoosten** [comment]
I can reproduce this issue, so let me start the analysis.

**Comment by attaxia** [comment]
@stefjoosten I think you misunderstood the issue. My issue is not that the Artikel interface does not allow editing, this is why I added the Artikeltype interface to begin with.

The issue is that from a clean install the Inleverplichtig checkbox should already be checked for artikeltype1, but it is not.

**Comment by attaxia** [comment]
I have slightly updated the example code and added a step to the reproduction path to avoid confusion.

**Comment by stefjoosten** [comment]
# Analysis

**Comment by stefjoosten** [comment]
@hanjoosten If this has to do with the tech-type `BOOLEAN`, I would expect this case to be tested in our regression. So in that case, I guess we should extend the regression set. But first we must make a diagnosis.

**Comment by attaxia** [comment]
Is it correct that this issue still has the wontfix label?

**Comment by stefjoosten** [comment]
That is incorrect. I have removed the wontfix label. It is however not likely that it will be fixed in the short term due to lack of time of all participants. (We could use help...)

**Comment by stefjoosten** [comment]
@attaxia ,
By the way, there are other ways to do what you want to do. Ways that make use of relations instead of booleans. Instead of a total, univalent relation `inleverPlichtig` between artikeltypes and booleans, you can define a relation `inleverPlichtig` between artikeltypes and the context in which these artikeltypes are inleverplichtig. In this relation you only put the artikeltypes that are inleverplichtig and NOT the ones that are not inleverplichtig.
E.g.
```Ampersand
RELATION inleverPlichtig [Artikeltype*ONE]

POPULATION inleverPlichtig CONTAINS
[ ("artikeltype1", ONE)
]
```
I have used ONE as something that defines the context, because there is nothing else in your script I can use for this purpose. In a more elaborate script, this might be (just as an example) an area, a shop, or whatever the context of the inleverplicht might be.

**Comment by RieksJ** [comment]
There is another way, which is by defining property-relations. Here is what you do:

First, you modify the relation that has a boolean:
~~~
--RELATION inleverPlichtig [Artikeltype*Inleverplichtig]  [UNI, TOT]
RELATION inleverPlichtig[Artikeltype] [PROP] -- This is a property relation
~~~
Property-relations are subsets of `I`, so populating them requires the following modification:
~~~
POPULATION inleverPlichtig CONTAINS
--[ ("artikeltype1", TRUE)
--, ("artikeltype2", FALSE)
[ ("artikeltype1", "artikeltype1") -- A populated pair is a 'true', an unpopulated pair is 'false'
]
~~~
The main difference between the use of property-relations and relations that use booleans is that the latter allow you to determine whether or not the boolean itself is defined. In your example, the construct `(I-inleverPlichtig;inleverPlichtig~)` tells you that the relation `inleverPlichtig` is not populated - nobody decided whether or not an Artikeltype is 'inleverplichtig'. You cannot do that with a property-relation: it is simply true or false, and never undefined.

So far, I have only very seldomly come across situations where I needed a BOOLEAN. I prefer property-relations because they are simpler, easier to use in expressions, and render nicely (as a checkbox) in user interfaces.

**Comment by Michiel-s** [comment]
My analysis of this issue:

1. When the Ampersand compiler generates the `population.json` file the initial population is correct
```json
{
    "links": [
        {
            "src": "artikeltype1",
            "tgt": "True"
        },
        {
            "src": "artikeltype2",
            "tgt": "False"
        }
    ],
    "relation": "inleverPlichtig[Artikeltype*Inleverplichtig]"
},
```
So far so good

2. When running the prototype installer, the database isn't populated correctly. As @stefjoosten already noticed:
![image](https://user-images.githubusercontent.com/7961582/108606079-4395ce80-73b8-11eb-915f-a194749dc2aa.png)

So here's the problem. The prototype framework doesn't process the boolean value correctly. Let's look into it

3. Check logs of prototype framework for population initialization
```logs
[2021-02-20 19:18:51] CORE.DEBUG: Add link (artikeltype1[Artikeltype],True[Inleverplichtig]) inleverPlichtig[Artikeltype*Inleverplichtig] to plug [] {"request_id":"aec49b6dc0"}
[2021-02-20 19:18:51] TRANSACTION.DEBUG: Mark relation 'inleverPlichtig[Artikeltype*Inleverplichtig]' as affected relation [] {"request_id":"aec49b6dc0"}
[2021-02-20 19:18:51] CORE.DEBUG: Atom artikeltype1[Artikeltype] already exists in concept [] {"request_id":"aec49b6dc0"}
[2021-02-20 19:18:51] CORE.DEBUG: Atom True[Inleverplichtig] already exists in concept [] {"request_id":"aec49b6dc0"}
[2021-02-20 19:18:51] DATABASE.DEBUG: UPDATE "Artikeltype" SET "inleverPlichtig" = '0' WHERE "Artikeltype" = 'artikeltype1' [] {"request_id":"aec49b6dc0"}
[2021-02-20 19:18:51] CORE.INFO: Link added to relation: (artikeltype1[Artikeltype],True[Inleverplichtig]) inleverPlichtig[Artikeltype*Inleverplichtig] [] {"request_id":"aec49b6dc0"}
```
Here's the problem. The population `(artikeltype1[Artikeltype],True[Inleverplichtig])` is wrongly translated to sql query `UPDATE "Artikeltype" SET "inleverPlichtig" = '0' WHERE "Artikeltype" = 'artikeltype1'`

4. So where in the code is this happening?
The method `Ampersand\Plugs\MysqlDB\MysqlDB::getDBRepresentation` in file [MysqlDB.php](https://github.com/AmpersandTarski/prototype/blob/development/src/Ampersand/Plugs/MysqlDB/MysqlDB.php) is responsible for this.
 
```php
  case "BOOLEAN":
    return (int) $atom->getId(); // booleans are stored as tinyint(1) in the database. false = 0, true = 1
```

The code wrongly casts to integer here, because $atom-getId() return the string "True" and not a boolean True. Casting string "True" to int in php returns `0`. Not `1`.

5. Solution
The following code replacement will fix it:
```php
case "BOOLEAN":
    // Booleans are stored as tinyint(1) in the database. False = 0, True = 1
    switch ($atom->getId()) {
        case "True":
            return 1;
        case "False":
            return 0;
        default:
            throw new Exception("Unsupported string boolean value '{$atom->getId()}'. Supported are 'True', 'False'", 500);
    }
    break;
```

**Comment by Michiel-s** [comment]
I'm transferring this issue to the Ampersand repo. There we keep track of all issues related to Ampersand compiler and prototype framework.
Thanks @attaxia for reporting this issue.
